File: longestrepeatingcharacterreplacement.py
from typing import List


class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        li: List[int] = [0] * 26
        left: int = 0
        max_fre: int = 0
        count: int = 0
        n: int = len(s)
        for right in range(n):
            li[ord(s[right]) - ord("A")] += 1
            max_fre = max(max_fre, li[ord(s[right]) - ord("A")])

            while (right - left + 1) - max_fre > k:
                li[ord(s[left]) - ord("A")] -= 1
                left += 1

            count = max(count, right - left + 1)

        return count


sol = Solution()
print(sol.characterReplacement("AABABBA", 1))

File: c++/longestrepeatingcharacterreplacement.cpp
#include <stdio.h>

using namespace std;
class Solution {
public:
  int characterReplacement(string s, int k) {
    // TC: O(n)
    int ans = 0;
    vector<int> freq(26, 0);
    int left = 0;
    int check = 0;
    for (int right = 0; right < s.size(); right++) {
      freq[s[right] - 'A']++;
      check = max(check, freq[s[right] - 'A']);
      while ((right - left + 1) - check > k) {
        freq[s[left] - 'A']--;
        left++;
      }
      ans = max(ans, right - left + 1);
    }

    return ans;
  }
};
